Automattic/wp-calypso
# E2E: Introduce TestAccount / Refactor LoginPage

This is a follow-up to https://github.com/Automattic/wp-calypso/pull/58516#discussion_r764400422.

### Suggested Changes

This PR is mainly about two things:

1. "Dumb down" the `LoginPage` model to remove (and prevent) the scope creep,
2. Introduce the `TestAccount` model to handle our e2e test accounts authentication.

In the spirit of keeping our `LoginPage` a simple POM, I came up with a concept of `TestAccount` model that would handle all the auth stuff related to our predefined E2E test accounts. In the first step, I created the new `LoginPage` from scratch, which was a great learning experience as it revealed, what I believe, some new good practices for our POMs - see the "Suggested good practices" section below. After many iterations of various account authentication concepts, I ended up creating the `TestAccount` model in the second step. It made sense to me to create this kind of high-level test account API as almost every suite (except account creation ones) revolves around our predefined test accounts. I think that from this place it would be best to let the code speak for itself, so please refer to the implementation for more details! 🙏 

Other changes:
- Make it possible to pass test account names via the `SAVE_AUTH_COOKIES` env var.
  This helps quickly pre-authenticate selected account(s) for faster local testing.
- Make the `setupHook` callback function async.
  This way it can be used also as `beforeAll` hook so we don't need to add an extra one.
- Remove the explicit `Log in` step and authenticate via `TestAccount` in the `setupHook`.
  We're authenticating some accounts via saved cookies, which is not a log-in step per se.
- Remove `clearAuthenticationState()` and `setLoginCookie()` from `BrowserManager`.
  This logic has been moved to the `TestAccount` model.
- Remove `startImmediately()` from the `NewPostFlow` as it's no longer used.
  As above, it's now handled by `TestAccount`. 
- Implement and use `gutenbergEditorPage.visit( 'post' )` instead of `newPostFlow.newPostFromNavbar()` to get to the "new post" page.
  See the "Define `visit()` method for each Page model" good practice suggestion below for the rationale on this one.
- Some general clean up.

### Testing instructions

- All tests should pass.
- Make sure default common accounts defined in the `globalSetup.ts` are pre-authenticated when the `SAVE_AUTH_COOKIES` var is set to `true`.
- Make sure pre-authentication is turned off when `SAVE_AUTH_COOKIES` var is set to `false`. 
- Make sure custom accounts are pre-authenticated when passed via the `SAVE_AUTH_COOKIES`, e.g.

  ```sh
  > export SAVE_AUTH_COOKIES=p2User,i18nUser
  > npx jest specs/specs-p2/p2-post.ts
  ```
- Make sure the authentication information is printed out when the `DEBUG` var is defined, e.g.

  ```sh
  > export DEBUG=pw:api
  > npx jest specs-p2/p2-post.ts
  ```
- Make sure the auth cookies are correctly considered `fresh` and `stale`.
  To test this one you can alter the 3 days value to just a few seconds in the `TestAccount.hasFreshCookies` method.

### Next steps

I've noticed that in some tests we use multiple accounts for the same browser context, for example [wp-support__showme-where.ts](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/specs-playwright/wp-support__showme-where.ts). This forces us to manually purge the current context in order to provide a fresh login state. The biggest disadvantage of this is that we cannot be sure we have a 100% separation this way, which we should. I think that with a fairly simple change we can create a new context for each `describe` block, which would let us free from worrying about any cookie leftovers, local states, or sessions. This would probably be good as a separate issue though - will create one!

---

### Suggested good practices for Page Model Objects based on the refactored LoginPage:

- **Define `visit()` method for each Page model.** 
  This helps quickly identify which page is handled by the current model. It also helps get directly to that page when needed instead of using flows to click the user through the navigation UI. I think that using navigation flows as a way of getting to the target page creates a risk of failure before even getting to the actual test subject. For example, we have a lot of editor test suites, each of which uses the Calypso interface to navigate to the editor page. Opening the editor via the UI should be tested separately, and for the actual editor tests, we can go to its page directly saving precious time and increasing test stability. As an example, this PR introduces also the `gutenbergEditorPage.visit()` method.
- **Don't redirect within page methods.**
  In other words, the redirect to the target page should always be called explicitly. Sometimes, the same page can have different URLs (like e.g. the log-in popup), and every POM method should be available for immediate use when the page is rendered, regardless of the current URL.
  As a supporting example, the actual log-in page sometimes opens as a popup window. If any of the LoginPage methods implied redirect (as e.g. `loginPage.logIn()` previously did), it would not be usable on the popped-up page window.  
- **Create small, single element/responsibility user actions.** 
  It makes the creation of the `selectors` object redundant and helps write readable flows where each step is self-described, e.g.
  
  ```js
  class LoginPage {
  // ...
  	async fillUsername( value: string ): Promise< Locator > {
  		const locator = await this.page.locator( 'input#usernameOrEmail' );
  		await locator.fill( value );
  	
  		return locator;
 	 }

  	async clickSubmit(): Promise< Locator > {
  		const locator = await this.page.locator( 'button[type="submit"]' );
  		await locator.click();

  		return locator;
  	}
  // etc.
  }
  ```
- **Use only existing element actions to create batch actions.**

  For example:

  ```js
  class LoginPage {
  // ...
  	async logIn( username, password ) {
  		await this.fillUsername( 'admin' );
  		await this.fillPassword( 'password' );
  		await Promise.all( [
  			this.page.waitForNavigation(),
  			this.clickSubmit(),
  		] );
  	}
  // etc.
  }
  ```
